### PR TITLE
Porting Guides 2.4

### PR DESCRIPTION
**Pull request by gundalow**
##### SUMMARY

Improve the user facing documentation to inform people what changes they may need to make when upgrading Ansible.

Porting guide for 2.4 only

**Note:** 2.0 and 2.3 porting guides have been moved to https://github.com/ansible/ansible/pull/27632

##### ISSUE TYPE
 - Docs Pull Request

##### COMPONENT NAME
docs

##### ANSIBLE VERSION
```
2.3.0
2.4.0
```

### CODE REVIEW

File: docs/docsite/rst/porting_guide_2.3.rst
Comment (calfonso) [inline docs/docsite/rst/porting_guide_2.3.rst]: Is this section needed for parity with previous versions of this document? If not, is it needed if it doesn't provide any info or than no info?

**Comment by thaumos** [comment]
I'd like to see this Guide in the sidebar, or somewhere in TOC.  Where did you think this doc would live in the tree?

File: docs/docsite/rst/porting_guide_2.0.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.0.rst]: Please avoid using 'etc'.

File: docs/docsite/rst/porting_guide_2.4.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.4.rst:49]: Please change to "...currently leads to the last tag specified overriding all of the previously specified tags."

File: docs/docsite/rst/porting_guide_2.4.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.4.rst]: Please revise to: "In 2.4, the default will be to merge the tags. You can enable the old overwriting behavior via the config option.

File: docs/docsite/rst/porting_guide_2.4.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.4.rst]: Please revise to: "The following modules will be removed in Ansible 2.5. Please update your playbooks accordingly."

File: docs/docsite/rst/porting_guide_2.4.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.4.rst]: "Noteworthy"

File: docs/docsite/rst/porting_guide_2.4.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.4.rst]: Please remove duplicate "number of changes to"

File: docs/docsite/rst/porting_guide_2.4.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.4.rst]: Add a period.

File: docs/docsite/rst/porting_guide_2.4.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.4.rst]: D the TBD

File: docs/docsite/rst/porting_guides.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guides.rst]: "This section lists porting guides that can help you upgrade your playbooks and modules from one version of Ansible to the next. Please note that this is not a complete list. If you believe any extra information would be useful in these pages, you can edit by clicking ` Edit on GitHub` on the top right, or raising an issue."

File: docs/docsite/rst/porting_guide_2.0.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.0.rst]: Please avoid using "etc". Suggest changing to either simply "playbooks and plugins" or maybe "your Ansible infrastructure".

File: docs/docsite/rst/porting_guide_2.0.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.0.rst]: Period after the word "updating", then a new sentence.

File: docs/docsite/rst/porting_guide_2.0.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.0.rst]: "playing tricks on Ansible" is ambiguous at best. Please revise and be specific.

File: docs/docsite/rst/porting_guide_2.0.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.0.rst]: Please remove "The ``fileglob`` lookup plugin is pretty simple to understand".

File: docs/docsite/rst/porting_guide_2.3.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.3.rst]: "This section discusses the behavioral changes between Ansible 2.2 and Ansible 2.3."

File: docs/docsite/rst/porting_guide_2.3.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.3.rst]: Please avoid using "etc". Suggest changing to either simply "playbooks and plugins" or maybe "your Ansible infrastructure".

File: docs/docsite/rst/porting_guide_2.3.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.3.rst]: "to understand what (if any) updates you will need to make."  Although removing the "(if any)" altogether might be better.

File: docs/docsite/rst/porting_guide_2.3.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.3.rst]: Please change to "...currently leads to the last tag specified overriding all of the previously specified tags."

File: docs/docsite/rst/porting_guide_2.3.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.3.rst]: Please revise to: "In 2.4, the default will be to merge the tags. You can enable the old overwriting behavior via the config option."

File: docs/docsite/rst/porting_guide_2.3.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.3.rst]: Please revise to "Here are some rare cases that might be encountered when updating. These are mostly caused by the more stringent parser validation and the capture of errors that were previously ignored."

File: docs/docsite/rst/porting_guide_2.3.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.3.rst]: Please revise to: "The following modules will be removed in Ansible 2.5. Please update your playbooks accordingly."

File: docs/docsite/rst/porting_guide_2.4.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.4.rst]: Please avoid using "etc". Suggest changing to either simply "playbooks and plugins" or maybe "your Ansible infrastructure".

File: docs/docsite/rst/porting_guide_2.4.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.4.rst]: Suggest changing to "...to understand what (if any) updates you will need to make." Although removing the "(if any)" altogether might be better.

File: docs/docsite/rst/porting_guide_2.4.rst
Comment (dharmabumstead) [inline docs/docsite/rst/porting_guide_2.4.rst]: perod after "porting", then new sentence "The complete list of porting guides can be found..."

**Review by dharmabumstead — CHANGES_REQUESTED** [review]
See edits. Thanks.

File: docs/docsite/rst/porting_guide_2.0.rst
Comment (gundalow) [reply docs/docsite/rst/porting_guide_2.0.rst]: Good spot. Not my words. Going to delete this sentence as it doesn't add any value

File: docs/docsite/rst/porting_guide_2.4.rst
Comment (gundalow) [reply docs/docsite/rst/porting_guide_2.4.rst]: How about `to understand what updates you may need to make.` to make it clear that it's possible that no updates will be required.

File: docs/docsite/rst/porting_guide_2.4.rst
Comment (gundalow) [reply docs/docsite/rst/porting_guide_2.4.rst:49]: To this?
> From now on, using ``--tags`` multiple times on one command line will emit a deprecation warning. Setting the ``merge_multiple_cli_tags`` option to True in the ``ansible.cfg`` file currently leads to the last tag specified overriding all of the previously specified tags.

File: docs/docsite/rst/porting_guide_2.3.rst
Comment (gundalow) [reply docs/docsite/rst/porting_guide_2.3.rst]: I wanted to keep the order roughly the same on the different pages, hence the headings.

However since there isn't anything here I've commented it out.

File: docs/docsite/rst/porting_guide_2.3.rst
Comment (gundalow) [reply docs/docsite/rst/porting_guide_2.3.rst]: Is this what you suggested?
Setting the ``merge_multiple_cli_tags`` option to True in the ``ansible.cfg`` file currently leads to the last tag specified overriding all of the previously specified tags.

**Comment by gundalow** [comment]
ready_for_review

I believe all comments have been addressed and would welcome a final review.

Thanks in advance

File: docs/docsite/rst/porting_guide_2.0.rst
Comment (twouters) [inline docs/docsite/rst/porting_guide_2.0.rst]: discusses? :-)

Comment (gundalow) [reply docs/docsite/rst/porting_guide_2.0.rst]: My fault for using a new RST editor that doesn't highlight typos, thanks.

File: docs/docsite/rst/porting_guide_2.3.rst
Comment (bcoca) [inline docs/docsite/rst/porting_guide_2.3.rst]: I would add actual example with full varname, i would move this to playbooks section above as these vars are normally used in plays

File: docs/docsite/rst/porting_guide_2.3.rst
Comment (bcoca) [inline docs/docsite/rst/porting_guide_2.3.rst]: `No major changes in this version` should be the default for each section that is empty

File: docs/docsite/rst/porting_guide_2.3.rst
Comment (bcoca) [inline docs/docsite/rst/porting_guide_2.3.rst]: I would remove oVirt/RHV this is 'new features' in module, unlike aws lambda and mount which change module behaviour and would require users to  reexamine existing playbooks.

File: docs/docsite/rst/porting_guide_2.3.rst
Comment (bcoca) [inline docs/docsite/rst/porting_guide_2.3.rst]: incorrect indentation and extra `-`

File: docs/docsite/rst/porting_guide_2.3.rst
Comment (bcoca) [inline docs/docsite/rst/porting_guide_2.3.rst]: Add the 'restructued async to work with action plugins' and the fact that this restores the usability of service + async

Comment (bcoca) [reply docs/docsite/rst/porting_guide_2.3.rst]: Add 'name' new to blocks .. so people can remove ugly` # commetn` on top of blocks

File: docs/docsite/rst/porting_guide_2.4.rst
Comment (bcoca) [inline docs/docsite/rst/porting_guide_2.4.rst]: missing:
- import_/include_ split
- ansible_facts namespacing
- inventory plugins: starting migration from hardcoded inventory + inventory scripts. scripts will still work via script plugin but efforts will now concentrate on plugins

File: docs/docsite/rst/porting_guide_2.4.rst
Comment (bcoca) [reply docs/docsite/rst/porting_guide_2.4.rst:49]: - add in 2.4 you can use multiple `-i /path/to/inventory`

File: docs/docsite/rst/porting_guide_2.4.rst
Comment (bcoca) [inline docs/docsite/rst/porting_guide_2.4.rst]: vars plugins have changed ... will fill in details on porting

File: docs/docsite/rst/porting_guide_2.4.rst
Comment (bcoca) [inline docs/docsite/rst/porting_guide_2.4.rst]: inventory scripts are becoming obsolete, in favor of inventory plugins

File: docs/docsite/rst/porting_guide_2.4.rst
Comment (bcoca) [inline docs/docsite/rst/porting_guide_2.4.rst]: deprecation cycle is +4 now so it would be 2.8

**Comment by gundalow** [comment]
Add
Python 2.4 and 2.5 support discontinuation

Ansible will not support Python 2.4 nor 2.5 on the target hosts anymore. Going forward, Python 2.6+ will be required on targets, as already is the case on the controller.

**Comment by gundalow** [comment]
Update the link in https://github.com/ansible/ansible/pull/27178/files

File: docs/docsite/rst/porting_guide_2.3.rst
Comment (gundalow) [reply docs/docsite/rst/porting_guide_2.3.rst]: 2.3 guide moved to other PR and has been merged. This issue has been addressed.

File: docs/docsite/rst/porting_guide_2.3.rst
Comment (gundalow) [reply docs/docsite/rst/porting_guide_2.3.rst]: 2.3 guide moved to other PR and has been merged. This issue has been addressed.

**Review by dharmabumstead — DISMISSED** [review]
Looks good to me, except for the remaining 'FIXME' tags. Please address those before merging. Thanks @gundalow!

File: docs/docsite/rst/porting_guide_2.4.rst
Comment (nitzmahone) [inline docs/docsite/rst/porting_guide_2.4.rst]: Don't know what format we're going for with this, but:
The `azure` module has been replaced by `azure_rm_virtualmachine`, which uses the new Resource Manager SDK.

Comment (gundalow) [reply docs/docsite/rst/porting_guide_2.4.rst]: Added in https://github.com/ansible/ansible/pull/24539/commits/33ab095f16a07d500f3509a6d87bef1a8b279691#diff-d11efc4cd1f939e08f3bd127476edb26R109

File: docs/docsite/rst/porting_guide_2.4.rst
Comment (ganeshrn) [inline docs/docsite/rst/porting_guide_2.4.rst]: Persistent config vars `connect_retries` and `connect_interval ` available in 2.3 is deprecated.
From `2.4` onward to set socket to connect timeout need to set variable `connect_retry_timeout` 

New timer added to control the command response timeout `command_timeout` in 2.4
Previously in 2.3 it was controlled using global timeout ie. `timeout` variable under `default`.

**Review by ganeshrn — APPROVED** [review]
Looks good for network section

**Comment by gundalow** [comment]
As we are very close to 2.4.0FINAL and in discussion with @abadger I've removed the FIXME sections and I'll create a follow up PR which @bcoca and others can update.

This will allow us to regenerate `docs.ansible.com/ansible` along side `2.4.0rc4`

**Comment by gundalow** [comment]
`FIXMEs` have been moved to https://github.com/ansible/ansible/pull/30456

**Comment by abadger** [comment]
hcerry picked for 2.4.0rc4

**Pull request merged**
